sip/sip-005-blocks-and-transactions.md

```diff
@@ -0,0 +1,1104 @@
+with otherwise-identically-named objects.
+
+When a key/value pair is inserted into the MARF, the hash of its key is
+calculated using the MARF's cryptographic hash function in order to determine
+where to insert the leaf.  The hash of the value is inserted as the leaf node,
+and the (hash, leaf) pair is inserted into the peer's data store.  This ensures
+that the peer can query any key/value pair on any fork it knows about in
+constant-time complexity.
+
+The text below describes the canonical encoding of key/value pairs that will be
+inserted into the MARF.
+
+#### Calculating a Fully-Qualified Object Name
+
+All objects' fully-qualified names start with the type of object they are,
+followed by an ASCII period `.`.  This can be the ASCII string "account",
+"smart-contract", "data-variable", or "data-map".
+
+Within an object type, a c32check-encoded addresses act as "namespaces" for keys in the state.  In all
+sets of key/value pairs, an ASCII period `.` is used to denote the separation between
+the c32check-encoded address and the following name.  Note that the c32 alphabet
+does _not_ include the ASCII period.
+
+#### Calculating a Typed Netstring
+
+A typed netstring is a variation of a
+[netstring](https://en.wikipedia.org/wiki/Netstring), which is a recursively-defined
+self-describing printable-ASCII encoding for type-carrying byte strings.
+Typed netstrings were chosen because each sequence of typed byte strings
+encodes to exactly one typed netstring.  While many typed variations of
+netstrings exist, the particular encoding used in Stacks is described here.
+
+Each type is encoded by a 1-byte alphabetical ASCII character as a prefix to the
+netstring length.  The following characters are used for Clarity types:
+
+* `i` for integer, which will be encoded as an ASCII string of numbers
+* `s` for buffer, which will be encoded as a lower-case hexadecimal string
+* `b` for boolean, which encodes to `1` for "true" or `0` for "false" 
+* `p` for standard principal, which will be encoded as an upper-case c32check string
+* `c` for contract principal, which will be encoded as an upper-case c32check string
+* `n` for a Clarity item name.  All Clarity items are encoded in ASCII already,
+  so converting them to a hex string is unnecessary.  In particular, this encoding is used for
+tuple names and asset names.
+* `t` for tuple.  Each item name is encoded as `n${len(name)}:${name},` followed
+  by its encoded value.  The name and value are in turn encoded as `t${total-len}:n${len(name)}:${name},${value_type}${len(value)}:${value},,`
+* `o` for option.
+   * `some x` is encoded as `o${len(x) + 4}:${x_type}${encode(x)},`.
+   * `none` is encoded as `o4:v0:,,`
+* `v` is a void type, used for `none` options.
+* `r` for result.
+   * `ok x` is encoded as `r${len(x) + 3}:o${x_type}${encode(x)},`
+   * `err x` is encoded as `r${len(x) + 3}:e${x_type}${encode(x)},`
+* `l` for list, which acts as a container for the above items.  Note that all
```

Comment:
I'd add a note that lists may themselves contain lists -- this would be representable in this netstring format like:

```
l27:l15:i1:1,i1:2,i2:-1,l5:i1:1
```

Also, since the netstring is explicitly specifiying lengths, terminating commas shouldn't be necessary.